### PR TITLE
fix(visx/annotation) Improve label positioning, add maxWidth

### PR DESCRIPTION
**Pull request by vovakulikov**
#### :boom: Breaking Changes
- Possible breaking change (remove default value of width at label component from @visx/annotation)

#### :bug: Bug Fix
- Fixes https://github.com/airbnb/visx/issues/1126

This PR is trying to improve horizontal label positioning at @visx/annotation package.  

Info from https://github.com/airbnb/visx/issues/1126

> Label component from @visx/annotation is trying to calculate right position for text within component according x,y,dx and dy props. But this component does’t take into account real width of text element

TLDR; This fix has added real width measures by `useMeasures` hook and now label component does calculation according real width and paddings value as well.
**Examples**

Problem - Bad label positioning example
https://codesandbox.io/s/bold-antonelli-0v17h?file=/src/pie-chart.tsx
GitHub repo branch https://github.com/vovakulikov/chart-grid-sandbox/tree/bad-label-positioning-example

**Example with fix** label position  https://codesandbox.io/s/focused-https-mvvjn
Github repo branch https://github.com/vovakulikov/chart-grid-sandbox/tree/label-positioning-with-annotation-fix

### CODE REVIEW

File: packages/visx-annotation/src/components/Label.tsx [hunk 1]
```diff
@@ -85,29 +85,35 @@ export default function Label({
   titleFontWeight = 600,
   titleProps,
   verticalAnchor: propsVerticalAnchor,
-  width = 125,
+  width: propWidth,
```
Comment (vovakulikov) [inline packages/visx-annotation/src/components/Label.tsx:88]: Possible breaking change.

File: packages/visx-annotation/src/components/Label.tsx [hunk 2]
```diff
@@ -85,29 +85,35 @@ export default function Label({
...
     padding.top + padding.bottom + (titleBounds.height ?? 0) + (subtitleBounds.height ?? 0),
   );
-  const innerWidth = width - padding.left - padding.right;
+
+  const measuredWidth = padding.right + padding.left + Math.max(
+    titleBounds.width ?? 0 ,
+    subtitleBounds.width ?? 0
+  );
+  const width = propWidth ?? measuredWidth;
+  const innerWidth = (width ?? measuredWidth) - padding.left - padding.right;
```
Comment (vovakulikov) [inline packages/visx-annotation/src/components/Label.tsx:109]: Fallback on real width of label in case if consumer did't pass label width

**Comment by vovakulikov** [comment]
@williaster can you help me with Happo UI tests. I don't have access to these tests therefore can't see what is going on there. Maybe I can fix that. Thanks.

**Review by williaster — COMMENTED** [review]
Hey @vovakulikov sorry for the delayed review, have been super busy. Thanks for another contribution! Happo has been a little flaky so don't worry about that / I just approved the diff. We're working on the issue, it basically has a stale base ref that it's comparing to for visual differences.

I think this approach works well for single-word labels as in your example. Otherwise there's a chicken-egg problem:
- the measured `width` is determined from whatever the `<Text />` component renders
- the `<Text />` component size is determined by `width` it is passed

So I think what happens if no `width` is specified initially
1) `measuredWidth = padding.left + padding.right ~= 24` (since measured = 0)
2) `measuredWidth=24` is passed to `Text`, which tries to render all of the text inside a `width` of `24`. `Text` won't break words, so a single word will render fine.
3) after `Text` is mounted/rendered, `measuredWidth` might update to include the new `measuredWidth` for whatever the (longest) word is

I pulled the branch to test it locally (the annotation example for happo specifies a width, so it wouldn't be affected by this change). This is what I get if I remove the `width`

image

If I dynamically update the text, you can see that the width updates to the longest word as described

Apr-08-2021 11-02-33

I think you **can** actually get the measured width to work *if you pass `undefined`* to `Text`, then it won't try to wrap based on a (non-existent) `width`. So in the case that `propsWidth` is undefined, I think we'd just pass `undefined` to the title + subtitle `Text` (and can probably remove all of the `measuredWidth` calculations since it'll be unused).

We've been trying to avoid breaking changes, I'll try to think on that a little bit. One thought I had was that even if you don't want to specify an explicit `width`, you might want auto-sizing with a `maxWidth`. So I wonder if we could do something like the following:
1) leave `width` as is, remove the default
2) add a `maxWidth` prop defaulted to the previous `width` default of `125`.

**Comment by vovakulikov** [comment]
@williaster Thanks for really comprehensive explanation about what is going on with width. 

It's getting start a really interesting bug because my example actually has a double-word title. I've checked my example again and added some suffixes just to check how it works and... 

Screenshot 2021-04-10 at 17 06 33

*(Screenshot from origin bug example with title suffix, **Notice** width for annotation was calculated according to length of title)*

So initial value for title was **calculated right for first render** but then we passed this first-calculated `measuredWidth` to `<Text/>` and by that we kind of lost our ability to measure width in a future because no matter what length of content we have, we have already passed width from first render to text. 

Actually this is almost the same that you've described above), So I totally agree to remove lines about passing width to `<Text />`.  

Screenshot 2021-04-10 at 17 29 44

*(Example without passing width to `<Text />`. Notice that we don't have word wrapping at all)*

But I can't figure out how to add this `maxWidth` prop to avoid this long text problem without using `useText` hook. It seems like we have to know some value of width before passing this value further to `<Text />`.

So as a result we have this visual behaviour

Screenshot 2021-04-10 at 18 34 48

*Labels with default `maxWidth` value 120px*

And gif about how it works at gallery example 
Screen Recording 2021-04-10 at 18 48 49

*Example from visx demo package, Notice in this example we have a small jump while text updating is going. Need investigate this further, right now don't have clear understanding why this is happening*

**A few words about this**
>and can probably remove all of the measuredWidth calculations since it'll be unused.

I've tried to remove this `measuredWidth` but it seems like we still need to have a width value since we use this to calculate adjusted coordinate for the container.

https://github.com/airbnb/visx/blob/42d3213fe5286388fc456eade9b3874d096616aa/packages/visx-annotation/src/components/Label.tsx#L111-L121

**Comment by vovakulikov** [comment]
Also https://github.com/airbnb/visx/issues/1123 probably related to useText calculation problem

**Comment by williaster** [comment]
Hey @vovakulikov thanks so much for the thorough follow up! I didn't quite have time to get to this today but will review again tomorrow 🙏

**Comment by felixfbecker** [comment]
@williaster reading the discussion/problem statements here, I wonder if using `<foreignObject>` with HTML elements would help/simplify in some way? Could that offer automatic text reflow?

**Comment by williaster** [comment]
@felixfbecker that's an interesting idea, we haven't really used that in `visx` at all. For `@visx/legend` and `@visx/tooltip` we opted to make them HTML based to get layout for free, and we have `@visx/text` for cases where you really probably should use SVG `<text />`s.

Annotations were initially implemented with SVG to get something out and usable, but we discussed having an HTML based variant as well at some point. `<foreignObject>` could be interesting to consider instead, though. maybe we can create a new issue to track that, since I think this change is still a strict win 🏆 .

File: packages/visx-annotation/src/components/Label.tsx [hunk 3]
```diff
@@ -100,6 +104,44 @@ export default function Label({
...
+  const subtitleMeasuredWidth = subtitleWordsByLine.reduce(
+    (maxSubtitleWidth, line) => Math.max(maxSubtitleWidth, line.width ?? 0),
+    0,
+  );
+
+  const textMeasuredWidth = Math.floor(
+    Math.min(maxWidth, Math.max(titleMeasuredWidth, subtitleMeasuredWidth)),
+  );
+  const measuredWidth = padding.right + padding.left + textMeasuredWidth;
+  const width = propWidth ?? measuredWidth;
```
Comment (williaster) [inline packages/visx-annotation/src/components/Label.tsx:144]: it's a bummer we can't skip the calculations above if `propWidth` **is** passed, but not easy to do with `hooks`. Theoretically could create a wrapper component explicitly to provide `width` if it's not passed, but I think we could consider that a future improvement if there are any perf issues. I didn't notice any lag in the annotation demo when quickly updating the title/subtitle text.

**Review by williaster — APPROVED** [review]
This looks great @vovakulikov, thanks again for the fix and addition of `maxWidth`! 

I pulled it locally to test it out as well and it seems pretty solid. I agree the edge cases in calculations likely come from the issue reported in #1123.

File: packages/visx-annotation/src/components/Label.tsx (hunk 3 above)
Comment (vovakulikov) [reply packages/visx-annotation/src/components/Label.tsx:144]: Totally with you about this. I like this idea about component wrapper with measure logic within.
